atoma-proxy/src/server/middleware.rs

```diff
@@ -1081,6 +1187,166 @@ pub mod auth {
+    /// * `user_id` - The ID of the user requesting the stack
+    /// * `request_model` - Implementation of RequestModel trait containing request details
+    /// * `endpoint` - The API endpoint being accessed
+    ///
+    /// # Returns
+    /// * `Result<SelectedNodeMetadata>` - Stack metadata if successful
+    ///
+    /// # Error Conditions
+    /// * Returns `AtomaProxyError::RequestError` if the maximum number of retry attempts is
+    ///   exceeded and the stack is still not available
+    ///
+    /// # Implementation Details
+    /// * Checks if the user's stack is locked using the users_buy_stack_lock_map
+    /// * If locked, retries up to MAX_STACK_WAIT_ATTEMPTS times with MAX_STACK_WAIT_TIME delay
+    /// * Each retry attempts to fetch the stack metadata via try_get_stack_for_user_id
+    /// * We don't wait for the Sui blockchain to finalize the stack creation, as we store the stack creation event
+    ///   right after the call to the Sui blockchain, and we don't wait to catch the event from the Sui blockchain,
+    ///   as this can take extra time (roughly 300ms following Sui's Mysticeti finality times).
+    #[instrument(level = "info", skip_all, fields(user_id = %user_id, endpoint = %endpoint), err)]
+    async fn get_stack_if_locked_with_request_model(
+        state: &ProxyState,
+        user_id: i64,
+        request_model: impl RequestModel + Send,
+        endpoint: &str,
+    ) -> Result<SelectedNodeMetadata> {
+        let stack_is_locked = {
+            state
+                .users_buy_stack_lock_map
+                .get(&user_id)
+                .is_some_and(|lock| *lock)
+        };
+        if stack_is_locked {
+            // NOTE: This means a concurrent request is already buying a stack, so we wait for it to finish,
+            // and for the stack creation event to be stored on the AtomaStateManager's internal state, as
+            // we store it right after the call to the Sui blockchain and don't wait to catch the event
+            // from the Sui blockchain, as this can take extra time (roughly 300ms following Sui's Mysticeti
+            // finality times). We will try again, for a fixed number of times. If the stack is still not created,
+            // we return an error.
+            for _ in 0..MAX_STACK_WAIT_ATTEMPTS {
+                let stack_metadata =
+                    try_get_stack_for_user_id(state, user_id, request_model.clone(), endpoint)
+                        .await?;
+                if let Some(stack_metadata) = stack_metadata {
+                    return Ok(stack_metadata);
+                }
+                tokio::time::sleep(MAX_STACK_WAIT_TIME).await;
+            }
+        }
+        Err(AtomaProxyError::RequestError {
+            message: format!("User {user_id} has already bought a stack"),
+            endpoint: endpoint.to_string(),
+        })
+    }
+
+    /// Arguments for the get_selected_node function.
+    pub struct GetSelectedNodeArgs<'a> {
```

Comment:
Why creating a struct to pass to a function, can't we just pass the arguments?

Reply:
Let's pass the struct, otherwise too many arguments, and I get lost passing them around in order